components/brave_sync/time_limited_words.cc

```diff
@@ -0,0 +1,172 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_sync/time_limited_words.h"
+
+#include <cmath>
+#include <string>
+#include <vector>
+
+#include "base/containers/span.h"
+#include "base/logging.h"
+#include "base/notreached.h"
+#include "base/strings/string_split.h"
+#include "base/strings/string_util.h"
+#include "base/time/time.h"
+#include "brave/components/brave_sync/crypto/crypto.h"
+#include "brave/components/brave_sync/qr_code_validator.h"
+#include "brave/vendor/bip39wally-core-native/include/wally_bip39.h"
+#include "brave/vendor/bip39wally-core-native/src/wordlist.h"
+
+namespace brave_sync {
+
+namespace {
+
+// TODO(alexeybarabash): subject to change
+static constexpr char kWordsv1SunsetDate[] = "Wed, 1 Jun 2022 00:00:00 GMT";
+
+}  // namespace
+
+using base::Time;
+using base::TimeDelta;
+
+Time TimeLimitedWords::words_v1_sunset_day_;
+
+std::string TimeLimitedWords::GetWordByIndex(size_t index) {
+  DCHECK_EQ(BIP39_WORDLIST_LEN, 2048);
+  index = index % BIP39_WORDLIST_LEN;
+  char* word = nullptr;
+  if (bip39_get_word(nullptr, index, &word) != WALLY_OK) {
+    LOG(ERROR) << "bip39_get_word failed for index " << index;
+    return std::string();
+  }
+
+  std::string str_word = word;
+  wally_free_string(word);
+
+  return str_word;
+}
+
+int TimeLimitedWords::GetIndexByWord(const std::string& word) {
+  std::string word_prepared = base::ToLowerASCII(word);
+
+  struct words* mnemonic_w = nullptr;
+  if (bip39_get_wordlist(nullptr, &mnemonic_w) != WALLY_OK) {
+    DCHECK(false);
+    return -1;
+  }
+
+  DCHECK_NE(mnemonic_w, nullptr);
+  size_t idx = wordlist_lookup_word(mnemonic_w, word_prepared.c_str());
+  if (!idx) {
+    return -1;
+  }
+
+  return idx - 1;
+}
+
+Time TimeLimitedWords::GetWordsV1SunsetDay() {
+  if (words_v1_sunset_day_.is_null()) {
+    bool convert_result =
+        Time::FromUTCString(kWordsv1SunsetDate, &words_v1_sunset_day_);
+    CHECK(convert_result);
+  }
+
+  CHECK(!words_v1_sunset_day_.is_null());
+
+  return words_v1_sunset_day_;
+}
+
+int TimeLimitedWords::GetRoundedDaysDiff(const Time& time1, const Time& time2) {
+  TimeDelta delta = time2 - time1;
+
+  double delta_in_days_f = delta.InMillisecondsF() / Time::kMillisecondsPerDay;
+
+  int days_rounded = std::round(delta_in_days_f);
+  return days_rounded;
+}
+
+std::string TimeLimitedWords::GenerateForNow(const std::string& pure_words) {
+  return TimeLimitedWords::GenerateForDate(pure_words, Time::Now());
+}
+
+std::string TimeLimitedWords::GenerateForDate(const std::string& pure_words,
+                                              const Time& not_after) {
+  int days_since_qrv2_epoh =
+      GetRoundedDaysDiff(QrCodeDataValidator::GetQRv1SunsetDay(), not_after);
```

Comment:
I think we need two hard-coded dates:
- The code V1 sunset date: a date in the **future** after which we don't accept the 24-word codes.
- The code V2 epoch date: a date in the **past** so that we can create valid codes starting now.

If I understand the implementation I'm looking at, right now it's not possible to generate valid codes because the `days_since_qrv2_epoch` will always be negative until June 1st.

Reply:
Thanks for the thorough review, @fmarier.

I use two dates:

1. as `epoch` I took the sunset date from the time-limited QR implementation which is `1 Jan 2022`;
2. as `sunset` I introduced `1 Jun 2022` which was expected the date when the PR migrates to the Stable channel.

I wanted to have `epoch` < `current_date` to have the actual working 25-words codephare while development. 
And then it would be possible to ensure on Nightly and Beta channels that it works in real life.

But I see that taking date from QR-code-related class confuses, so I am redoing this place. Time-limited words epoch will be `15th of April`

Reply:
thanks, I fixed this at https://github.com/brave/brave-core/pull/12300/commits/4c59158b9070eacf74afa52c42c1930b85f942e4

Reply:
Thanks for the clarifications @AlexeyBarabash . These changes makes this part a lot clearer!